File: src/knowledge-store.ts
import { knowledgeGraphPath } from "./paths.js";
import { readJson, writeJson } from "./files.js";
import type { KnowledgeNode, TailQuestion } from "./types.js";

interface KnowledgeGraph {
  nodes: KnowledgeNode[];
  version: number;
}

export async function loadKnowledgeGraph(): Promise<KnowledgeGraph> {
  try {
    return await readJson<KnowledgeGraph>(knowledgeGraphPath);
  } catch (error) {
    // If it doesn't exist, return empty graph structure
    return { nodes: [], version: 1 };
  }
}

export async function saveKnowledgeGraph(graph: KnowledgeGraph): Promise<void> {
  await writeJson(knowledgeGraphPath, graph);
}

export async function addKnowledgeNode(node: KnowledgeNode): Promise<void> {
  const graph = await loadKnowledgeGraph();
  
  // Filter out existing node if any to avoid duplication
  graph.nodes = graph.nodes.filter(n => n.articleId !== node.articleId);
  graph.nodes.push(node);
  
  await saveKnowledgeGraph(graph);
}

export async function getTodos(status?: 'todo' | 'in_progress' | 'done'): Promise<{ articleId: string; topic: string; todo: TailQuestion }[]> {
  const graph = await loadKnowledgeGraph();
  const todos: { articleId: string; topic: string; todo: TailQuestion }[] = [];
  
  for (const node of graph.nodes) {
    for (const q of node.tailQuestions) {
      if (!status || q.status === status) {
        todos.push({
          articleId: node.articleId,
          topic: node.topic,
          todo: q
        });
      }
    }
  }
  return todos;
}

export async function updateTodoStatus(
  questionId: string,
  status: 'todo' | 'in_progress' | 'done',
  linkedArticleId?: string
): Promise<boolean> {
  const graph = await loadKnowledgeGraph();
  let updated = false;
  
  for (const node of graph.nodes) {
    for (const q of node.tailQuestions) {
      if (q.id === questionId) {
        q.status = status;
        if (linkedArticleId) {
          q.linkedArticleId = linkedArticleId;
        }
        updated = true;
      }
    }
  }
  
  if (updated) {
    await saveKnowledgeGraph(graph);
  }
  return updated;
}

/**
 * Calculates backlinks between a new article and existing nodes in the knowledge graph.
 * For example, if the new article's topic or keyword matches a tail question in an existing article,
 * we can create a backlink from the existing article to the new one.
 */
export async function calculateBacklinks(
  newArticleId: string,
  newTopic: string
): Promise<{ fromArticleId: string; toArticleId: string; anchor: string }[]> {
  const graph = await loadKnowledgeGraph();
  const backlinks: { fromArticleId: string; toArticleId: string; anchor: string }[] = [];
  
  // A simple keywords/topic matching rule:
  // If an existing article has a tail question (TODO) whose words or concepts overlap with the newTopic,
  // we link them.
  const newTopicNormalized = newTopic.toLowerCase().replace(/\s+/g, "");
  
  for (const node of graph.nodes) {
    if (node.articleId === newArticleId) continue;
    
    for (const q of node.tailQuestions) {
      const qNormalized = q.question.toLowerCase().replace(/\s+/g, "");
      
      // If the question contains key parts of the new topic, suggest a backlink
      // (For a production system this could be semantic search or LLM based, but let's do keyword matching for now)
      const isMatch = qNormalized.includes(newTopicNormalized) || newTopicNormalized.includes(qNormalized);
      if (isMatch) {
        backlinks.push({
          fromArticleId: node.articleId,
          toArticleId: newArticleId,
          anchor: q.question
        });
        
        // Auto-update the tail question to point to this new article
        q.status = 'done';
        q.linkedArticleId = newArticleId;
      }
    }
  }
  
  return backlinks;
}

File: src/pipeline/knowledge_store.py
import re
from typing import List, Dict, Any, Optional
from src.core.paths import project_root
from src.core.files import read_json, write_json
from src.core.types import KnowledgeNode, TailQuestion

knowledge_graph_path = project_root / "content" / "knowledge-graph.json"

def load_knowledge_graph() -> Dict[str, Any]:
    try:
        return read_json(knowledge_graph_path)
    except Exception:
        return {"nodes": [], "version": 1}

def save_knowledge_graph(graph: Dict[str, Any]) -> None:
    write_json(knowledge_graph_path, graph)

def add_knowledge_node(node: KnowledgeNode) -> None:
    graph = load_knowledge_graph()
    node_data = node.model_dump(exclude_none=True)
    
    # Filter out node with same articleId to prevent duplication
    graph["nodes"] = [n for n in graph.get("nodes", []) if n.get("articleId") != node.articleId]
    graph["nodes"].append(node_data)
    
    save_knowledge_graph(graph)

def get_todos(status: Optional[str] = None) -> List[Dict[str, Any]]:
    graph = load_knowledge_graph()
    todos = []
    
    for node in graph.get("nodes", []):
        article_id = node.get("articleId", "")
        topic = node.get("topic", "")
        for q in node.get("tailQuestions", []):
            if not status or q.get("status") == status:
                todos.append({
                    "articleId": article_id,
                    "topic": topic,
                    "todo": TailQuestion.model_validate(q)
                })
    return todos

def update_todo_status(
    question_id: str,
    status: str,
    linked_article_id: Optional[str] = None
) -> bool:
    graph = load_knowledge_graph()
    updated = False
    
    for node in graph.get("nodes", []):
        for q in node.get("tailQuestions", []):
            if q.get("id") == question_id:
                q["status"] = status
                if linked_article_id:
                    q["linkedArticleId"] = linked_article_id
                updated = True
                
    if updated:
        save_knowledge_graph(graph)
    return updated

def calculate_backlinks(
    new_article_id: str,
    new_topic: str
) -> List[Dict[str, str]]:
    graph = load_knowledge_graph()
    backlinks = []
    
    new_topic_normalized = re.sub(r"\s+", "", new_topic.lower())
    
    for node in graph.get("nodes", []):
        if node.get("articleId") == new_article_id:
            continue
            
        for q in node.get("tailQuestions", []):
            q_normalized = re.sub(r"\s+", "", q.get("question", "").lower())
            
            # Simple keyword overlap match
            is_match = (new_topic_normalized in q_normalized) or (q_normalized in new_topic_normalized)
            if is_match:
                backlinks.append({
                    "fromArticleId": node.get("articleId"),
                    "toArticleId": new_article_id,
                    "anchor": q.get("question")
                })
                
                # Auto-update status to done and link the new article
                q["status"] = "done"
                q["linkedArticleId"] = new_article_id
                
    if backlinks:
        save_knowledge_graph(graph)
        
    return backlinks
